### ISSUE TITLE
Amethyst causing OS crash on latest version

### ISSUE THREAD

**Issue by danpalmer** [ianyh/Amethyst#988]
**Describe the bug**
I believe Amethyst's most recent version is causing an OS crash. Not quite sure if it's a "kernel panic", but has the same effect. I believe it's Amethyst as a) it's the only app I have that I know messes about with windowing, b) it's an always-running utility, and c) I know it's recently had an update.

When opening my laptop having been away for some time, say overnight, I find the laptop is no longer on, and boots up. On startup I get a message from the OS saying my computer was rebooted because of a problem. This problem appears to be that the WindowServer crashed and/or did not respond to a sort of ping from the OS for a period of time.

A suspicious crash report in Console shows the WindowServer crashing at around the time I closed my laptop for the evening.

I've now experienced this crash 3 times in the last 4-5 days. It doesn't seem to happen every time, but regularly.

**Applications:**
It appears to be an interaction with the WindowServer.

**To Reproduce**

Nothing concrete, but leaving Amethyst running while closing a laptop and coming back to it some time later.

**Expected behavior**
Amethyst should not hard-crash the OS.

**Versions:**
 - macOS: 10.15.3
 - Amethyst: 0.15.1 (85)
 
**Debug Info**
```
$ /Applications/Amethyst.app/Contents/MacOS/Amethyst --debug-info
Version: 0.15.1 (85)

OS version: Version 10.15.3 (Build 19D76)

Has permissions: true

Screens:
	(0.0, 0.0, 1440.0, 2560.0) [(0.0, 0.0, 1440.0, 2560.0)]
	(-1440.0, 812.0, 1440.0, 900.0) [(-1440.0, 848.0, 1440.0, 900.0)]
	(1440.0, 0.0, 1440.0, 2560.0) [(1440.0, 0.0, 1440.0, 2560.0)]

Manageable applications:
	ViewBridgeAuxiliary (com.apple.ViewBridgeAuxiliary)
	AppSSOAgent (com.apple.AppSSOAgent)
	talagent (com.apple.talagent)
	ViewBridgeAuxiliary (com.apple.ViewBridgeAuxiliary)
	Control Strip (com.apple.controlstrip)
	Google Chrome (com.google.Chrome)
	Code (com.microsoft.VSCode)
	GitHub Desktop (com.github.GitHubClient)
	iTerm2 (com.googlecode.iterm2)
	Calendar (com.apple.iCal)
	Notes (com.apple.Notes)
	Things3 (com.culturedcode.ThingsMac)
	Messages (com.apple.iChat)
	Music (com.apple.Music)
	Keynote (com.apple.iWork.Keynote)
	Soulver (com.acqualia.soulver)
	Preview (com.apple.Preview)
	Dash (com.kapeli.dashdoc)
	QuickLookUIService (PID 457) (com.apple.quicklook.QuickLookUIService)
	Finder (com.apple.finder)
	EmojiFunctionRowIM (com.apple.inputmethod.EmojiFunctionRowItem)
	Legacy Color Picker Extensions (Keynote) (com.apple.appkit.xpc.LegacyExternalColorPickerService)
	CalNCService (com.apple.CalendarNotification.CalNCService)
	1Password 7 (com.agilebits.onepassword7)
	ToothFairy (com.robinlu.mac.Tooth-Fairy)
	SiriNCService (com.apple.SiriNCService)
	studentd (com.apple.studentd)
	Console (com.apple.Console)
	Simulator (com.apple.CoreSimulator.SimulatorTrampoline)
	com.apple.CoreSimulator.CoreSimulatorService (com.apple.CoreSimulator.CoreSimulatorService)

Configuration:
window-margins: 0
screen-padding-top: 0
focus-follows-mouse: 1
window-resize-step: 5
restore-layouts-on-launch: 1
follow-space-thrown-windows: 1
screen-padding-left: 0
mod2: (
    option,
    shift,
    control
)
window-minimum-height: 0
layouts: (
    tall,
    fullscreen,
    column,
    bsp,
    row
)
floating: (
        {
        id = "com.apple.systempreferences";
        "window-titles" =         (
        );
    }
)
screen-padding-bottom: 0
mod1: (
    option,
    shift
)
debug-layout-info: 0
new-windows-to-main: 0
mouse-follows-focus: 0
enables-layout-hud: 1
ignore-menu-bar: 0
screen-padding-right: 0
use-canary-build: 0
enables-layout-hud-on-space-change: 0
mouse-swaps-windows: 1
window-minimum-width: 0
window-margin-size: 7
floating-is-blacklist: 1
float-small-windows: 1
```

**Additional context**
Attached a crash report for the WindowServer.

[WindowServer_2020-03-16-194135_danmbp.crash.zip](https://github.com/ianyh/Amethyst/files/4342525/WindowServer_2020-03-16-194135_danmbp.crash.zip)

**Comment by AndreiBarsan** [comment]
Yeah, I've been getting similar crashes since using Amethyst as well. I've attached a crash report.

[WindowServer_2020-04-11-140534_andreib-C02Y71MJJGH6.crash.zip](https://github.com/ianyh/Amethyst/files/4475544/WindowServer_2020-04-11-140534_andreib-C02Y71MJJGH6.crash.zip)

**Comment by AndreiBarsan** [comment]
I think this may be correlated with some laggy behavior in Amethyst. Sometimes, my Amethyst commands can take multiple seconds (occasionally >10s) to actually execute, during which the Mac menu bar is completely frozen (beach ball). 

I am running the following programs on my mac (Catalina 10.15.4):

- Emacs 26.3
- Google Chrome
- Safari
- VSCode
- iTerm2
- Zoom
- Toggl
- Slack
- Preview
- Activity Monitor

**Comment by AndreiBarsan** [comment]
This froze again this morning, with over >1min lag in responding to certain commands like window resizing. It *seems* to be related to Pulse Secure's VPN credential prompt (which triggers every ~24h to renew my work VPN connection), but the correlation isn't very consistent.

I did not have such freezes before starting to use Amethyst, when I was just using Pulse Secure.

I looked at Activity Monitor and neither Pulse nor Amethyst were responsive. Killing Amethyst did not unfreeze my system, but killing Pulse and re-starting both did.

My version of Amethyst is 0.15.3 (88).

**Comment by kit-ty-kate** [comment]
I think I’ve just hit this issue with 0.15.4 on the M1 macbook air (arm64 binary):

<img width="1061" alt="Screenshot 2021-07-14 at 11 00 10" src="https://user-images.githubusercontent.com/2611789/125607367-b43d5309-4239-46ec-824d-ea157f41ee05.png">

**Comment by ianyh** [comment]
This kinda feels like a macOS kernel bug. We don't do anything at the level of the window server. Can anyone verify if this still happens on macOS 15 (or 26)?
